sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/PasswordPlaintextFastHashingCheck.kt

```diff
@@ -0,0 +1,127 @@
+    qualifier = "javax.crypto.SecretKeyFactory"
+    name = "getInstance"
+    withArguments(STRING_TYPE)
+}
+private val pbeKeySpecConstructorMatchers = listOf(
+    FunMatcher {
+        qualifier = "javax.crypto.spec.PBEKeySpec"
+        matchConstructor = true
+        withArguments(CHAR_ARRAY_TYPE, BYTE_ARRAY_TYPE, INT_TYPE)
+    },
+    FunMatcher {
+        qualifier = "javax.crypto.spec.PBEKeySpec"
+        matchConstructor = true
+        withArguments(CHAR_ARRAY_TYPE, BYTE_ARRAY_TYPE, INT_TYPE, INT_TYPE)
+    },
+)
+
+@Rule(key = "S5344")
+class PasswordPlaintextFastHashingCheck : CallAbstractCheck() {
+
+    override val functionsToVisit = listOf(generateSecretFunMatcher)
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: KaFunctionCall<*>,
+        kotlinFileContext: KotlinFileContext,
+    ) = withKaSession {
+        generateSecretFunMatcher.matches(resolvedCall) || return@withKaSession
+        val secretKeyFactoryCall = callExpression.predictReceiverExpression()
+            ?.resolveToCall()
+            ?.successfulFunctionCallOrNull()
+            ?.takeIf { getInstanceFunMatcher.matches(it) }
+            ?: return@withKaSession
+        val keySpecCall = resolvedCall.argumentMapping.keys.single().predictRuntimeValueExpression()
+            .resolveToCall()
+            ?.successfulConstructorCallOrNull()
+            ?.takeIf { call -> pbeKeySpecConstructorMatchers.any { matcher -> matcher.matches(call) } }
+            ?: return@withKaSession
+
+        val iterationCountExpression = keySpecCall.argumentMapping.keys.elementAtOrNull(2) ?: return@withKaSession
+        val iterationCountValueExpression = iterationCountExpression
+            .predictRuntimeValueFromParameterDefault()
+            .predictRuntimeValueExpression()
+        val iterationCount = iterationCountValueExpression.predictRuntimeIntValue() ?: return@withKaSession
+
+        val algorithmValueExpression = secretKeyFactoryCall.argumentMapping.keys.single()
+            .predictRuntimeValueFromParameterDefault()
+            .predictRuntimeValueExpression()
+        val algorithm = algorithmValueExpression.stringValue() ?: return@withKaSession
+        val minIteration = minIterations[algorithm] ?: return@withKaSession
+
+        if (iterationCount < minIteration) {
+            val secondaryLocations = mutableListOf(SecondaryLocation(kotlinFileContext.textRange(algorithmValueExpression)))
+            if (iterationCountValueExpression.textRange != iterationCountExpression.textRange) {
+                secondaryLocations.add(SecondaryLocation(kotlinFileContext.textRange(iterationCountValueExpression)))
+            }
+
+            kotlinFileContext.reportIssue(iterationCountExpression, MESSAGE.format(minIteration), secondaryLocations)
+        }
+    }
+
+    private fun KtExpression.predictRuntimeValueFromParameterDefault(): KtExpression = withKaSession {
```

Comment:
I was thinking of extracting this method into general utilities, but its use is very specific and very contextual, and in general could make false prediction since non-val parameters can be reassigned before their use, within the method.

So in the end I left it as a `private fun`.